### ISSUE TITLE
Allow additional Tags for activity creation

### ISSUE THREAD

**Issue by martinjt** [dotnet/aspnetcore#50488]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

When doing sampling with Activity/Traces, the information that you have to make the decision on whether to Sample of Drop is only that which is passed in as tags to Start/CreateActivity method. Sampling at the Head (root activity) is the most efficient method as it allows child activitys to inherit that decision and therefore reduce processing. You can sample at an individual (child) activity level too.

Right now, there is very little information to go on when trying to apply Head Sampling in a .NET application. This is because of the Activity being created having no tags being used as part of the Create.

https://github.com/dotnet/aspnetcore/blob/2f7a3f73bdf916eee798182f3e804bfdda1a57cd/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L389

This means that the samplers can only base their decision on the parent context.

In order to use other contextual information, you need to use something like the IHttpContextAccessor to access the ambient context.

### Describe the solution you'd like

What would make this better from a consumer perspective would be an extension point to augment the activity tags at creation time.

I don't know if Features for HTTP are available to get during the pipeline, however, something like the following

```csharp
interface IHttpActivityEnrichmentFeature
{
   ActivityTagsCollection OnCreate(HttpContext context);
   void PostCreate(Activity activity, HttpContext context);
}
```

This would allow users to do that enrichment at the right time. It would allow the level of data being used to be configured according to the performance concerns of the user.

**Comment by joegoldman2** [comment]
What do you think about renaming the two methods to `ActivityCreating` and `ActivityCreated`?

```cs
/// <summary>
/// Defines a feature to enrich activities with additional tags.
/// </summary>
public interface IHttpActivityEnrichmentFeature
{
    /// <summary>
    /// Called during the creation of an activity.
    /// </summary>
    /// <param name="context">The current HTTP context.</param>
    ActivityTagsCollection ActivityCreating(HttpContext context);

    /// <summary>
    /// Called after an activity has been created to perform further enrichment.
    /// </summary>
    /// <param name="activity">The created activity.</param>
    /// <param name="context">The current HTTP context.</param>
    void ActivityCreated(Activity activity, HttpContext context);
}
```

This convention is already used in some places like https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Diagnostics/IDbConnectionInterceptor.cs.

**Comment by JamesNK** [comment]
Latest update: https://github.com/dotnet/aspnetcore/pull/62090#issuecomment-3191076170

tldr we'll specify the recommended OTEL semantic conventions HTTP span attributes on activity creation. This aligns with our goal to put the HTTP span attributes on the activity (post .NET 10)

**Comment by rkargMsft** [comment]
10.0-rc1 is out now:

> add AppContext.SetSwitch("Microsoft.AspNetCore.Hosting.SuppressActivityOpenTelemetryData", false) to your ASP.NET Core app startup code. Later in .NET 11 we would fully support OTEL HTTP spec. It would be enabled by default and you could remove the switch.
